### ISSUE TITLE
Crash with C-Locale in non-ASCII PWD when path is looked up

### ISSUE THREAD

**Issue by faho** [fish-shell/fish-shell#2614]
As mentioned in https://github.com/fish-shell/fish-shell/pull/2381#issuecomment-164884889, I get a crash when I:

- cd to a path with non-ASCII characters in it (`mkdir /tmp/ü; cd /tmp/ü`)

- Launch fish with LC_ALL set to C (`set -gx LC_ALL C; fish`)

- Type `c` (which will trigger a `cd` suggestion)

```fish
$ mkdir /tmp/ü
$ cd /tmp/ü
$ set -gx LANG C
$ fish
>fish: Could not determine current working directory. Is your locale set correctly?
>fish: src/path.cpp:148: bool path_get_cdpath(const wcstring&, wcstring*, const wchar_t*, const env_vars_snapshot_t&): Assertion `wd[len - 1] == L'/'' failed.
>fish: 'fish' terminated by signal SIGABRT (Abort)
```

Stubbing out the assert makes it not crash, though it doesn't quite behave _great_ - the PWD isn't shown fully and suggestions seem to eat into the prompt - every "ü" seems to move the cursor one character to the left, removing the character that's there.

**Comment by zanchey** [comment]
I'd be interested to see what other shells do in this situation, because as far as I can tell this is undefined territory.

**Comment by faho** [comment]
bash 4.3.42 seems to just work, including working $PWD - I can cd out and in of the directory, which includes entering it (which also doesn't work with fish). This is also with no locale variable except LC_ALL (set to C).

**Comment by krader1961** [comment]
That bash seems to handle this doesn't surprise me. I'm betting it's not decoding the UTF-8 sequences found in the path name to unicode code points when reading the PWD. And if it simply emits that raw UTF-8 sequence to a terminal that expects UTF-8 then the correct characters will be displayed. I would, however, expect bash to be confused about the amount of horizontal space consumed by the PWD but that might be hard to notice.

This is going to take some careful thought to fix properly. Merely reading a config file like /etc/locale.conf and hoping a UTF-8 locale is thereby defined is a bandaid, not a comprehensive solution. I'm still adamantly opposed to adding the reading and parsing of any such file to the core fish code. But a (hopefully) temporary hack that works well enough most of the time would be to could hardcode setting __LC_ALL__ to __en_US.UTF-8__ early in the fish bootstrap sequence if neither it or __LANG__ was set. That would at least keep fish from barfing on a non-ASCII PWD. That might buy us some time to implement a proper fix that doesn't involve a bunch of distro specific hacks.

P.S., I also think there are real bugs in ArchLinux and probably other distros in this regard. If there is a "system" level locale that should already be present in the process environment when fish is started. Fish should not have to know about /etc/locale.conf or any equivalent file.

**Comment by krader1961** [comment]
From my most recent comment on PR #2381 which attempts to "fix" this problem in a distro specific manner:

> Also, read http://unix.stackexchange.com/questions/39175/understanding-unix-file-name-encoding for a pretty good discussion of the issue regarding shells like fish handling non-ASCII path names.

See also my [comment on that pull request](https://github.com/fish-shell/fish-shell/pull/2381#issuecomment-167461798) about ISO 8859 and the futility of trying to accommodate it.

**Comment by krader1961** [comment]
This specific issue was fixed by my commit c2f1df1d4af0c7e633528cb4c8caa79ef04b0b5a to make fish behave more sanely in the C/POSIX locales.
